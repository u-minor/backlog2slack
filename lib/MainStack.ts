import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { HostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';
import { CloudFrontResource } from './CloudFrontResource';
import { DynamoDBResource } from './DynamoDBResource';
import { LambdaResource } from './LambdaResource';
import { Route53Resource } from './Route53Resource';

interface Props extends StackProps {
  certificate: ICertificate;
}

export class MainStack extends Stack {
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props);

    const domainName = this.node.tryGetContext('domainName');
    const hostName = this.node.tryGetContext('hostName');

    const dynamoDBResource = new DynamoDBResource(this);

    const lambdaResource = new LambdaResource(this, {
      backlog: {
        apiKey: this.node.tryGetContext('backlogApiKey'),
        host: this.node.tryGetContext('backlogHost'),
      },
      dynamoDB: { table: dynamoDBResource.table },
      slack: {
        botToken: this.node.tryGetContext('slackBotToken'),
        signingSecret: this.node.tryGetContext('slackSigningSecret'),
      },
    });

    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', {
      domainName: domainName,
    });

    const cloudfront = new CloudFrontResource(this, {
      certificate: props.certificate,
      domainName: `${hostName}.${domainName}`,
      lambda: { functionUrl: lambdaResource.functionUrl.url },
    });

    new Route53Resource(this, {
      hostedZone,
      recordName: `${hostName}.${domainName}`,
      target: RecordTarget.fromAlias(
        new CloudFrontTarget(cloudfront.distribution),
      ),
    });

    new CfnOutput(this, 'Endpoint URL', {
      description: 'Endpoint URL of backlog2slack',
      value: `https://${hostName}.${domainName}/`,
    });
  }
}
