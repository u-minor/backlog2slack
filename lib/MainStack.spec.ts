import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { HostedZone } from 'aws-cdk-lib/aws-route53';
import { MainStack } from './MainStack';

describe('Backlog2SlackStack', () => {
  let template: Template;

  beforeAll(() => {
    const certificate = Certificate.fromCertificateArn(
      new cdk.Stack(),
      'TestCertificate',
      'arn:aws:acm:us-east-1:1234567890:certificate/DUMMY_ID',
    );
    jest.spyOn(HostedZone, 'fromLookup').mockReturnValue({
      hostedZoneId: 'MY_ZONE_ID',
      zoneName: 'example.com',
    } as unknown as HostedZone);

    const app = new cdk.App({
      context: {
        domainName: 'example.com',
        hostName: 'test',
      },
    });
    const stack = new MainStack(app, 'TestStack', {
      certificate,
      env: {
        account: '1234567890',
        region: 'us-east-1',
      },
    });

    template = Template.fromStack(stack);
  });

  it('should generate valid output', () => {
    const actual = template.findOutputs('*');

    expect(actual).toMatchSnapshot();
  });
});
