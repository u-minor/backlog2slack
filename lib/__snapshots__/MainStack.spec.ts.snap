// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backlog2SlackStack should generate valid output 1`] = `
{
  "EndpointURL": {
    "Description": "Endpoint URL of backlog2slack",
    "Value": "https://test.example.com/",
  },
}
`;
